presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java

```diff
@@ -331,10 +331,20 @@ public PlanWithProperties visitAggregation(AggregationNode node, StreamPreferred
             PlanWithProperties child = planAndEnforce(node.getSource(), childRequirements, childRequirements);
 
             List<VariableReferenceExpression> preGroupedSymbols = ImmutableList.of();
-            if (!LocalProperties.match(child.getProperties().getLocalProperties(), LocalProperties.grouped(groupingKeys)).get(0).isPresent()) {
+            // SortedByKeys here is already the longest prefix of the sorted-by columns that is a subset of the grouped-by columns
+            List<LocalProperty<VariableReferenceExpression>> sortedByKeys = child.getProperties().getLocalProperties();
+            if (!LocalProperties.match(sortedByKeys, LocalProperties.grouped(groupingKeys)).get(0).isPresent()) {
                 // !isPresent() indicates the property was satisfied completely
                 preGroupedSymbols = groupingKeys;
             }
+            else {
+                // Enable segmented aggregation if the prefix of the sorted-by keys is a subset of the grouped-by keys
+                // Use the longest prefix to maximize the segment aggregation's preGroupedSymbols
+                List<Optional<LocalProperty<VariableReferenceExpression>>> matchResult = LocalProperties.match(sortedByKeys, LocalProperties.grouped(groupingKeys));
```

Comment:
This match function is really confusing. We should improve its function comments. :)
What does it return if columns in groupingKeys have a different order from sortedByKeys?